staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/validating/dispatcher.go

```diff
@@ -62,30 +62,54 @@ func newValidatingDispatcher(p *Plugin) func(cm *webhookutil.ClientManager) gene
 	}
 }
 
+var _ generic.VersionedAttributeAccessor = &versionedAttributeAccessor{}
+
+type versionedAttributeAccessor struct {
+	versionedAttrs map[schema.GroupVersionKind]*admission.VersionedAttributes
+}
+
+func (v *versionedAttributeAccessor) VersionedAttribute(attr admission.Attributes, o admission.ObjectInterfaces, gvk schema.GroupVersionKind) (*admission.VersionedAttributes, error) {
+	if val, ok := v.versionedAttrs[gvk]; ok {
+		return val, nil
+	}
+	versionedAttr, err := admission.NewVersionedAttributes(attr, gvk, o)
+	if err != nil {
+		return nil, err
+	}
+	v.versionedAttrs[gvk] = versionedAttr
+	return versionedAttr, nil
+}
+
+func (v *versionedAttributeAccessor) VersionedAttributeFromCache(gvk schema.GroupVersionKind) (*admission.VersionedAttributes, bool) {
+	if val, ok := v.versionedAttrs[gvk]; ok {
+		return val, true
+	} else {
+		return nil, false
+	}
+}
+
 var _ generic.Dispatcher = &validatingDispatcher{}
 
 func (d *validatingDispatcher) Dispatch(ctx context.Context, attr admission.Attributes, o admission.ObjectInterfaces, hooks []webhook.WebhookAccessor) error {
 	var relevantHooks []*generic.WebhookInvocation
 	// Construct all the versions we need to call our webhooks
-	versionedAttrs := map[schema.GroupVersionKind]*admission.VersionedAttributes{}
+	versionedAttrAccessor := &versionedAttributeAccessor{
+		versionedAttrs: map[schema.GroupVersionKind]*admission.VersionedAttributes{},
+	}
 	for _, hook := range hooks {
-		invocation, statusError := d.plugin.ShouldCallHook(hook, attr, o)
+		invocation, statusError := d.plugin.ShouldCallHook(ctx, hook, attr, o, versionedAttrAccessor)
 		if statusError != nil {
 			return statusError
 		}
 		if invocation == nil {
 			continue
 		}
+
 		relevantHooks = append(relevantHooks, invocation)
-		// If we already have this version, continue
-		if _, ok := versionedAttrs[invocation.Kind]; ok {
-			continue
-		}
-		versionedAttr, err := admission.NewVersionedAttributes(attr, invocation.Kind, o)
+		_, err := versionedAttrAccessor.VersionedAttribute(attr, o, invocation.Kind)
```

Comment:
Add a comment that this is caching the result here, so it can be reused in parallel later.

Reply:
done